ha/util/consul_kv_store.py

```diff
@@ -0,0 +1,174 @@
+
+        Args:
+            prefix (str): Consul prefix
+            host (str): consul host
+            port (str): consul port
+
+        Example:
+            ConsulKvStore("cortx/ha/system/config", host="consul.srv", port=3000)
+            ConsulKvStore("cortx/ha/system/config")
+        Here localhost:8500 is connection and cortx/ha/system/config is prefix.
+        """
+        self._prefix: str = prefix
+        self._verify(prefix, host, port)
+        self._consul = self._get_connection(prefix, host, port)
+        self._consul.kv.put(self._prepare_key(""), None)
+
+    def _verify(self, prefix: str, host: str, port: int):
+        """
+        Verify connection detail.
+
+        Args:
+            prefix (str): Consul prefix
+            host (str): consul host
+            port (str): consul port
+        """
+        if prefix == "":
+            raise Exception("Invalid prefix. It cannot be empty.")
+        if host != "localhost":
+            try:
+                socket.gethostbyname(host)
+                int(port)
+            except Exception as e:
+                raise Exception(f"Invalid host or port name. Refused connection to {host}. Error: {e}")
+
+    def _get_connection(self, prefix: str, host: str, port: int):
+        """
+        Prepare consul connection.
+
+        Args:
+            prefix (str): Consul prefix
+            host (str): consul host
+            port (str): consul port
+
+        Return:
+            Object: Consul object.
+        """
+        if host=="localhost" and port==8500:
+            return consul.Consul()
+        return consul.Consul(host=host, port=port)
+
+    def _verify_data(self, *args):
+        """
+        Verify data.
+        """
+        for data in args:
+            if data is None or data == "":
+                raise Exception("Invalid parameter. Data cannot be none or empty.")
+
+    def _prepare_key(self, key):
+        """
+        Prepare key with prefix.
+
+        Args:
+            key (str): Get Key.
+        """
+        key: list = [x for x in self._prefix.split("/") + key.split("/") if x != ""]
+        return "/".join(key)
+
+    def key_exists(self, key: str):
+        """
+        Check if key exists.
+
+        Args:
+            key (str): Consul Key.
+
+        Return:
+            bool: True if key exists else False.
+        """
+        _, data = self._consul.kv.get(self._prepare_key(key), recurse=True)
+        if data is None:
+            return False
+        return True
+
+    def set(self, key: str, val: str=None):
+        """
+        Set key-val pair in consul. If key already exists return exception.
+
+        Args:
+            key (str): Key.
+            val (str): Value.
+
+        Return:
+            str: Return value.
+        """
+        self._verify_data(key)
+        if self.key_exists(key):
+            raise Exception(f"Key {key} already exists in kv store.")
+        self._consul.kv.put(self._prepare_key(key), val)
```

Comment:
It would be good to have blocking and non-blocking variants of get/set/update functions in this class. The user can choose which one to use. This class is a common code and should be in cortx-util. Currently it implements basic functionality to unblock 3-node ha provisioning. We can have proper functionality implemented when we move this code to utils repo. @ajay-paratmandali please add this suggestion in the class.

Reply:
yes, we need some retries on certain errors, we had to do this in SSPL earlier.

Reply:
Added TODO at top of class

Reply:
Why not to implement it in scope of current issue? The case i've described is not  a floating bug - it is quite a common situation which happened often on R1 cluster.

Reply:
I tried `@repeat_if_fails` this and look easy and blocking call till connection is back.

Yes, I think both blocking and nonblocking call should be available.
From HA service may need blocking call but resource may need non blocking

We can have
```
get/set/update/delete
get_block/set_block/update_block/delete_block
```
There is one limitation here, **as it may move in cortx-util, there is no log provided, So it is very difficult** to check if service getting halt somewhere.

@ajaysrivas  @indrajitzagade  Please share your opinion 
how should we proceed

Reply:
It will be implemented later. Let's unblock 3-node mini-provisioner.